pkg/resource/deploy/step_generator.go

```diff
@@ -883,13 +858,127 @@ func (sg *stepGenerator) continueStepsFromRefresh(event ContinueResourceRefreshE
 			sg.deployment.newPlans.set(urn, newResourcePlan)
 		}
 
+		cts := &promise.CompletionSource[*resource.State]{}
+		// Set up the cts to trigger a continueStepsFromImport when it resolves
+		go func() {
+			// if promise had an "ContinueWith" like method to run code after a promise resolved we'd use it here,
+			// but a goroutine blocked on Result and then posting to a channel is very cheap.
+			old, err := cts.Promise().Result(context.Background())
+
+			// Create a fresh 'new' state which will be for the final goal state, the import step will have 'mutated'
+			// the original 'new' variable. We still need that mutated state for the display and snapshot layers to
+			// reference, but we need a new clean separate state for the the step generator and any follow up steps to
+			// use.
+			var new *resource.State
+			if err == nil {
+				new = resource.NewState(
+					goal.Type, urn, goal.Custom, false, "", goal.Properties, nil, goal.Parent, protectState, false,
+					goal.Dependencies, goal.InitErrors, goal.Provider, goal.PropertyDependencies, false,
+					goal.AdditionalSecretOutputs, aliasUrns, &goal.CustomTimeouts, "", retainOnDelete, goal.DeletedWith,
+					createdAt, modifiedAt, goal.SourcePosition, goal.IgnoreChanges, goal.ReplaceOnChanges)
+			}
+
+			sg.events <- &continueResourceImportEvent{
+				RegisterResourceEvent: event,
+				err:                   err,
+				urn:                   urn,
+				old:                   old,
+				new:                   new,
+				provider:              prov,
+				invalid:               event.Invalid(),
+				recreating:            recreating,
+				randomSeed:            randomSeed,
+			}
+		}()
+
+		sg.imports[urn] = true
 		if isReplace := old != nil && !recreating; isReplace {
 			return []Step{
-				NewImportReplacementStep(sg.deployment, event, old, new, goal.IgnoreChanges, randomSeed),
+				NewImportReplacementStep(sg.deployment, event, old, new, goal.IgnoreChanges, randomSeed, cts),
 				NewReplaceStep(sg.deployment, old, new, nil, nil, nil, true),
-			}, false, nil
+			}, true, nil
+		}
+		return []Step{NewImportStep(sg.deployment, event, new, goal.IgnoreChanges, randomSeed, cts)}, true, nil
```

Comment:
any races we need to worry about at this interleaving point? I don't have any offhand, but I continue to feel a bit uneasy about the possibilities with async step generation.

Reply:
I don't think so, we've already saved the URN to stop duplicates. I don't think this adds any more raciness to the system.